### ISSUE TITLE
Too easy to close the window by mistake

### ISSUE THREAD

**Issue by rpwoodbu** [rpwoodbu/mosh-chrome#5]
Right now, if you hit the "close" button, the session is just gone without warning. Happily, CTRL-W does not close the window, but is passed to the terminal. Nonetheless, this is problematic.

There are two obvious solutions to this:
1. Keep the session going in the background. This would require some UI changes most likely. The session can always be totally destroyed with `Ctrl-^ .`. This would be especially nice in conjunction with multiple simultaneous sessions (issue #4).
2. Pop up an "are you sure" dialog. Relatively easy to do, and probably worth doing, even if we take it out later in favor of another solution.

**Comment by rpwoodbu** [comment]
Although I knew this (but didn't say it explicitly in this issue), it was pointed out in Issue #107 that closing the window forcibly closes NaCl, preventing it from signaling the server that the session should be torn down. This may be surprising to users.

I (or someone) should revisit the assertion I made about Option 2 being a non-starter; this would be very easily solved if I could capture the window close intent.

Alternatively, I could move the NaCl embed into the background page. As I said before, this would complicate communications a bit, but it seems tractable. It is more invasive a chance than I would like just to fix this issue; I would much rather find a solution to Option 2.

**Comment by acornejo** [comment]
any ideas on how to achieve this? I keep closing the window by accident, when I open it again it creates a new mosh server, I end up with multiple mosh servers running and this is a single user machine,... so then I have to go clean the pids myself. Its annoying.

**Comment by rpwoodbu** [comment]
All the ideas I have are captured already in this issue. I don't have time to invest in this, but ideas and especially pull requests are welcome. It'd be great if there were a simple way to achieve Option 2 (or a variant of it).

**Comment by acornejo** [comment]
Option 1 is definitely feasible, but I would not know how to implement this with rewriting large portions of this plugin. I am new to chrome apps, but it seems that if we put the nacl part on the background page then we would have to open a socket between hterm and the background page to allow them to communicate.

Option 2 seems hard due to chrome packaged apps restrictions as you pointed out.

Let me propose an alternative:

Option 3: Register a listener in the background page for the window close event on the main app, we can't prevent it but we can at least know about. Then when this event is triggered we open a new window with a popup that offers the user to reconnect to the session just closed. This would require  some modifications on mosh to allow a new client to connect (it seems we just need to pass whatever nonce/sequence number the client had before it exited to the background app).

**Comment by rpwoodbu** [comment]
Yes, Option 1 would require some work to implement, but the communication isn't really the hard part. There's already communication happening between NaCl and a mosh_window page just to get stuff into and out of the terminal emulator (which is in JavaScript). I'd have to look into it to be sure, but it may be easy to direct that same communication method to an arbitrary page in the app. The main reasons I haven't tried is that I don't find myself accidentally closing windows that often, and I really want a more general solution (see below).

Option 2 may be hard, but maybe not if someone knows a "trick". One thing that has occurred to me is that I've noticed the occasional app which has replaced the top bar with something of their own choosing. It seems it would be possible to replace it with something devoid of a "close" button (or at least one that is much smaller, or out of the way). It would still be possible to close the window forcibly by right-clicking on the icon in the panel, but it would be unlikely to happen by accident.

Regarding Option 3: I've wanted a way to "revive" a stranded session for as long as I've used Mosh. If we could do that, there would be lots benefits beyond just resolving this issue, including automatically restoring sessions between client reboots! The official story from the upstream Mosh project has been that they're not trying to reinvent screen/tmux, and don't want to introduce the potential security vulnerability of "hijacking" a session. I have a few things to say about that. First, there's a ton of overlap between Mosh and screen/tmux already (e.g., all are complete terminal emulators). Second, this is a user flow issue that Mosh can solve easily, but which the combination of Mosh and screen/tmux makes more complicated (and requires you to (remember to) use screen/tmux). Third, this is not a major security vulnerability; the way I imagine it, it would still require that you login via ssh in order to reset the session sync'd state and establish a new key, and this would be done through a secured named pipe, similar to how ssh-agent communication works.

So I'm all about Option 3 if you can make it happen!

**Comment by rpwoodbu** [comment]
I just realized that I glossed over your parenthetical re Option 3. I actually tried to do exactly what you described regarding the nonce, but it doesn't work because there's no way to reset the server's idea of what state the client has. So you'd have to persist the entirety of the terminal state. That just seems like a bridge too far. I did consider trying to add a "state reset" command to the protocol, but the more I thought about it, the more I realized that this is not the winning solution.

**Comment by acornejo** [comment]
Thanks for taking the time to reply.

I had forgotten that this is a stateful protocol, so your analysis seems spot on, option 3 will be impossible without at least some customization of mosh upstream that allows the client to send a reset signal to the server.

Regarding option 2, do you know of any such app? I am new to chrome os, but as far as I can tell every app has exactly the same top bar, given that this operating system touts security I doubt there is a way to customize it. If you know an exception to this rule let me know and if its open source (most chrome apps are) I can take a peak and see if I can port the behavior to mosh.

PS: I share your frustration about resuming sessions, and the annoyance of dead/stale mosh-servers running when connecting to new clients. For the lack of a better solution upstream I've resorted to the following script:

```bash
#!/bin/bash
SESSION_NAME=${1:-0}
if tmux list-sessions | grep ${SESSION_NAME} &>/dev/null; then
    exec tmux attach -d -t ${SESSION_NAME}
else
    exec tmux new -s ${SESSION_NAME}
fi
```

By using that script as the remote command to run with mosh, it will attach to any existing tmux session if it exists, or create a new one if none exists. Since it uses the -d option on attach, and it runs with exec (no bash intermediary process) it means that when any old session is detached the stale mosh server days. This means at any one time I only have one mosh-server running for a given tmux session.

**Comment by rpwoodbu** [comment]
The [Chrome Remote Desktop](https://chrome.google.com/webstore/detail/chrome-remote-desktop/gbchcmhmhahfdphkhkmpfmihenigjmpp?utm_source=chrome-app-launcher-search) app has a bar which looks similar in many ways to the OS-provided bar, but adds a full-screen button and a hamburger menu.

Just the other day I was looking for a decent serial terminal emulator (I could find a _decent_ one; I should just hack one together frontended by hterm), and tried [Nuwa](https://chrome.google.com/webstore/detail/nuwa-serial-port-debug-to/ccgeahdeedajelmcfjednddijkfccipe?utm_source=chrome-app-launcher-search). It's kinda terrible, but it does demonstrate a totally custom top bar. And interestingly, half the time the "close" button doesn't even work... which actually is a hint that this is possible!

(I have a similar script for cleaning up stale Mosh sessions by naming each client uniquely (like you do with `$SESSION_NAME`), and could hack it to reconnect to tmux as well. These are all manifestations of the same problem, which could be neatly solved as a first-class feature of Mosh.)

**Comment by acornejo** [comment]
Ahh!

This is great, thank you. I think you've hit the nail in the head, it seems the option we want is readily available. Namely `frame: 'none'` as the create window options will give us a frameless window. This is in fact properly documented:

https://developer.chrome.com/apps/app_window#type-CreateWindowOptions

Using that option would allow the app to create its own frame with minimize/maximize/close buttons whose actions are completely customizable and permit a confirmation dialog.

**Comment by rpwoodbu** [comment]
Thanks for running that down! This would enable a few other niceties (some of which have been requested in the past), such as:
 * Showing the connected machine in the bar.
 * Hiding the bar for extra vertical real estate.
 * Adding a discoverable menu (that wouldn't conflict with hterm's context menu) for features such as:
   * Creating a new session.
   * Switching sessions.

I would want the new bar to be as similar as possible to the stock one in look and function. In particular, ChromeOS has a few tricks up its sleeve, such as dragging on the maximize button to "dock" the window. Not sure how much work it would be to keep that behavior.
